text/0002-local-frames-revision.md

```diff
@@ -0,0 +1,195 @@
+
+`MAV_FRAME_BODY_NED` description makes clear that it's frame with body orientation (fixed to vehicle): _"e.g. useful to command 2 m/s^2 acceleration to the right"_, however it's name implies it's NED frame, which is confusing. Description of `MAV_FRAME_BODY_OFFSET_NED` makes thing even more complicated because it looks like it's a real NED frame: _"e.g. useful to command 2 m/s^2 acceleration to the east"_.
+
+The only detailed explanation of current local frames meaning is in APM documentation [[5]](http://ardupilot.org/dev/docs/copter-commands-in-guided-mode.html#set-position-target-local-ned). It describes `MAV_FRAME_LOCAL_OFFSET_NED` as vehicle-carried NED frame and `MAV_FRAME_BODY_OFFSET_NED` as vehicle-carried Forward-Right-Down frame which is clear. However, behavior of `MAV_FRAME_BODY_NED` frame is complicated:
+
+_Positions are relative to the vehicle’s home position in the North, East, Down (NED) frame. Use this to specify a position x metres north, y metres east and (-) z metres above the home position. Velocity directions are relative to the current vehicle heading. Use this to specify the speed forward, right and down (or the opposite if you use negative values)._
+
+So the behavior of frames for positions and for velocities is not always the same, which is confusing and should be avoided. The naming pattern is unclear:
+
+* it uses three fields (`LOCAL` or `BODY`, `OFFSET` or not, `NED` or `ENU`) to describe in fact two entities: frame of reference translation (origin) and rotation. Futhermore, rotation is somehow defined by combination of `LOCAL`/`BODY` and `NED`/`ENU` which is excessive;
+* `BODY` frames are not always aligned with respect to the vehicle orientation;
+* `NED` doesn't really mean anything except direction of z axis, e.g. `MAV_FRAME_BODY_OFFSET_NED` is not NED-oriented.
+
+Both `_OFFSET_` frames are not supported by PX4 firmware at all, while `MAV_FRAME_BODY_NED` is supported only for velocity setpoints (where it acts as Forward-Right-Down oriented frame, not NED).
+
+Furthermore, `MAV_FRAME` enumeration lacks a frame which is fixed in orientation to the moving vehicle. Most similar frame, `MAV_FRAME_BODY_OFFSET_NED`, have only the yaw fixed to the vehicle.
+
+# Detailed Design
+
+## Frames of reference naming
+
+The proposed frames naming convention is:
+
+`MAV_FRAME_origin_rotation[_modifier]`,
+
+where `origin` must be one of the following:
+
+* `LOCAL` for frames with local origin fixed with respect to the Earth;
+* `BODY` for frames with origin fixed to the vehicle;
+
+`rotation` must be one of the following:
+
+* `NED` for North-East-Down;
+* `ENU` for East-North-Up;
+* `FRD` for Forward-Right-Down, where Forward and Right are tangential to the Earth and Down is normal to it;
+* `FLU` for Forward-Left-Up, where Forward and Left are tangential to the Earth and Up is normal to it;
+* `RPY` for Roll-Pitch-Yaw (fixed in orientation to the moving vehicle, z axis points down when vehicle is leveled, right-handed);
```

Comment:
Isn't this the same as the BODY_FRD?

Reply:
In `FRD` frames only yaw is fixed to the vehicle, and z axis always stays normal to the Earth (vertical). In other words, `FRD` frame is the `NED` frame, rotated about the z axis by the vehicle yaw angle. In `RPY` all three axes are fixed to the vehicle, i.e. x axis is the vehicle roll axis. So in `RPY` frame z axis is not normal to the ground when the vehicle is tilted (not level).